src/Microsoft.AspNetCore.SystemWebAdapters/Authentication/RemoteAppAuthenticationModule.Framework.cs

```diff
@@ -2,47 +2,112 @@
+
+    public void Init(HttpApplication context)
+    {
+        context.PostMapRequestHandler += (s, _) =>
+        {
+            var context = ((HttpApplication)s).Context;
             if (string.Equals(context.Request.Path, _authOptions.AuthenticationEndpointPath, StringComparison.OrdinalIgnoreCase)
                 && context.Request.HttpMethod.Equals("GET", StringComparison.OrdinalIgnoreCase))
             {
-                if (!string.Equals(_remoteAppOptions.ApiKey, context.Request.Headers.Get(_remoteAppOptions.ApiKeyHeader), StringComparison.Ordinal))
+                MapRemoteAuthenticationHandler(new HttpContextWrapper(context));
+
+                if (context.Handler is null)
                 {
-                    // Using 407 here (proxy authentication required) to differentiate from the scenario of
-                    // a valid API key but no authenticated user.
-                    context.Response.StatusCode = 407;
                     context.ApplicationInstance.CompleteRequest();
                 }
-                else
-                {
-                    context.Handler = handler;
-                }
             }
+        };
+    }
+
+    public void MapRemoteAuthenticationHandler(HttpContextBase context)
+    {
+        var apiKey = context.Request.Headers.Get(_remoteAppOptions.ApiKeyHeader);
+        var migrationAuthenticateHeader = context.Request.Headers.Get(AuthenticationConstants.MigrationAuthenticateRequestHeaderName);
+
+        if (migrationAuthenticateHeader is null)
+        {
+            // If no migration authentication header is present, then this request was not initiated by the ASP.NET Core app
+            // for authentication purposes (though, of course, the request likely did proxy through that app).
+            //
+            // This most likely indicates that an identity provider is redirecting back to the application after
+            // authenticating the user. In that case, the original-url query string will indicate the path
+            // that the user should be redirected back to.
+            var originalUrlPath = context.Request.QueryString[AuthenticationConstants.OriginalUrlQueryParamName];
+            if (originalUrlPath is not null)
+            {
+                context.Response.StatusCode = 302;
+
+                // Redirect back to the provider path relative to the current host
+                var redirect = new Uri(context.Request.Url, originalUrlPath);
+                context.Response.Headers["Location"] = redirect.ToString();
+            }
+            else
+            {
+                // A request without a migration authentication header and without an original URL to redirect
+                // back to is invalid. Return 400 to indicate that the request was malformed.
+                context.Response.StatusCode = 400;
+            }
+
+            // Clear any existing handler as this request is now completely handled
+            context.Handler = null;
```

Comment:
It's not necessarily null. This is run post mapping request handlers and I noticed while testing the scenario that this was typically set to an MVC handler (so if this module doesn't change anything, the MVC handler will attempt to handle the request).

There isn't a controller for /systemweb-adapters/authenticate, of course, so the MVC handler was overwriting the 302 or 400 or whatever that this module set with a 401.

I think this may not have been necessary previously when the module always set the handler for requests to the authenticate path. But now that the module is doing some of the checks for whether to handle the request or not, we need to clear the handler in cases that we aren't explicitly setting it (and don't want to fall back to whatever the default would be like MVC).